Implemented screen_get_usable_rect for X11 Display Server

pouleyKetchoupp opened the pull request:
Helps with fixing issues with scrolling popups not respecting screen decorations on the display server side for Linux.

~Extra work is still needed for `PopupMenu` to handle usable screen size properly for all platforms (see #41564).~
*edit:* this has been fixed in #41640

**Edit - Support for multi-screen:**
~Handling decorations is only partially supported, because there's no proper mechanism available in X11 to get accurate information about each screen separately on all desktop systems. There are two cases:~
- ~`_GTK_WORKAREAS` is used when available from the WM to get accurate rect for the different screens (available on Gnome).~
- ~Otherwise `_NET_WORKAREA` is used. It provides one full rect for all screens, which doesn't handle decorations on the secondary screen in all cases (KDE, XFCE).~

**Edit2 - Update for multi-screen:**
Handling decorations is supported in different ways depending on the information the window manager provides:
- `_GTK_WORKAREAS` is used when available from the WM to get accurate rect for the different screens directly (available on Gnome).
- Alternatively, strut information is used to calculate available space for a given desktop manually (XFCE, KDE).
- As last resort `_NET_WORKAREA` is used. It provides one full rect for all screens, which doesn't handle decorations on the secondary screen in all cases.

pouleyKetchoupp committed:
Implemented screen_get_usable_rect for X11 Display Server

Helps with fixing issues with scrolling popups not respecting screen
decorations on the display server side.

Reproduction steps for a simple use case:
- Start the editor project list
- Open the language selection popup

Support for multi-screen:
Handling decorations is supported in different ways depending on the
information the window manager provides:
- _GTK_WORKAREAS is used when available from the WM to get accurate rect
for the different screens directly (available on Gnome).
- Alternatively, strut information is used to calculate available space
for a given desktop manually (XFCE, KDE).
- As last resort _NET_WORKAREA is used. It provides one full rect for all
screens, which doesn't handle decorations on the secondary screen in all
cases.

pouleyKetchoupp commented:
Just pushed an update to handle multiple screens. However it has only partial support for handling decorations properly on some configurations.

ssokolow commented:
I only just learned about Godot being limited to `_NET_WORKAREA` on non-GNOME desktops. 

This isn't just a problem because of how it interacts with panels... it also reacts quite badly if your desktop is made of differently-sized monitors, because `_NET_WORKAREA` has to take the largest rectangle which will fit entirely inside your desktop, and that'll cut off non-panel regions of some monitors.

I had to solve this for my [QuickTile](http://ssokolow.com/quicktile/) project and can advise you on how to go about it. (Or, if it would help, I think I'm the sole author of the relevant Python code. If you'd like me to confirm that and discuss letting you at it under something more permissive than the GPL to use as reference, I'm open to that.)

The gist of it is that, outside GNOME, you have to reinvent the code the WM uses to generate `_NET_WORKAREA` and its internal equivalent to `_GTK_WORKAREAS`:
1. Use `XGetWindowProperty` to read the [`_NET_CLIENT_LIST`](https://specifications.freedesktop.org/wm-spec/1.3/ar01s03.html) property (a list of top-level windows maintained by the window manager) on the root window.
2. For the root window and each window ID listed in `_NET_CLIENT_LIST`, use `XGetWindowProperty` to query `_NET_WM_STRUT_PARTIAL`, and, if it's unset, `_NET_WM_STRUT` as a fallback.
2. [Parse](https://specifications.freedesktop.org/wm-spec/1.3/ar01s05.html) the strut definitions thus received.
3. Subtract them from the usable desktop.

(Ironically, QuickTile currently has an open bug for not working properly under GNOME. GNOME Shell doesn't bother doing the courtesy of setting `_NET_WM_STRUT_PARTIAL` or `_NET_WM_STRUT` for its own panels, and my life's been a mess since around the time I became aware that I needed to support `_GTK_WORKAREAS`.)

For my code, I wanted to properly support using the space adjacent to panels that don't span a full edge of the screen, so I didn't go for the "generate a per-monitor `_NET_WORKAREA` solution. Instead, my code generateds a candidate rectangle for the window and then moves or resizes it in whichever of the four cardinal directions would require the least absolute movement to move it off the reserved rectangle. (`Rectangle.moved_off_of` and `Rectangle.subtract` in my utility classes.)

pouleyKetchoupp commented:
@ssokolow Thanks so much for your comment and all the details! It seems like it would help a lot indeed, I'll give it a try when I can find some time.

Also, in case you're interested to contribute, don't hesitate to make a commit with a better implementation for non-gnome desktops. I'd be happy to integrate it to this PR and co-author you.

ssokolow commented:
> Also, in case you're interested to contribute, don't hesitate to make a commit with a better implementation for non-gnome desktops. I'd be happy to integrate it to this PR and add you as co-author.

I wish I had the time. As-is, I'm still struggling to catch up my existing obligations enough to give some TLC to some of my own projects that are in dire need of it.

pouleyKetchoupp commented:
Rebased and updated to handle multi-screen decorations properly on XFCE & KDE.

Thank you so much @ssokolow for your help! I was able to make it work correctly using strut information from your instructions. I'm handling edge cases by checking for the intersection size and it seems to work fine in all the cases I've tested.

@akien-mga Let's wait for a few days to see if someone can review the code, otherwise we can do the same as for #42117 and merge so we can start getting feedback in the alpha.

ssokolow commented:
> Thank you so much @ssokolow for your help! I was able to make it work correctly using strut information from your instructions. I'm handling edge cases by checking for the intersection size and it seems to work fine in all the cases I've tested.

I'm glad I could help. :)

akien-mga commented:
Thanks!
